xtra-libp2p/src/endpoint.rs

```diff
@@ -526,26 +529,53 @@ impl Endpoint {
             "Type-system enforces that we only try to negotiate one protocol"
         );
 
-        let (protocol, stream) = self.open_substream(peer, protocols.clone()).await?;
+        let (control, _) = self.controls.get(&peer).ok_or(Error::NoConnection(peer))?;
 
-        debug_assert!(
-            protocol == protocols[0],
-            "If negotiation is successful, must have selected the only protocol we sent."
-        );
+        let fut = {
+            let connection_timeout = self.connection_timeout;
+            let control = control.clone();
+            async move {
+                let (protocol, stream) =
+                    Self::open_substream(control, peer, protocols.clone(), connection_timeout)
+                        .await?;
 
-        Ok(stream)
+                debug_assert!(
+                    protocol == protocols[0],
+                    "If negotiation is successful, must have selected the only protocol we sent."
+                );
+
+                Ok(stream)
+            }
+        };
+
+        Ok(Box::pin(fut))
     }
 
+    #[must_use]
     async fn handle(
         &mut self,
         msg: OpenSubstream<Multiple>,
-    ) -> Result<(&'static str, Substream), Error> {
+    ) -> Result<
+        Pin<Box<dyn futures::Future<Output = Result<(&'static str, Substream), Error>> + Send>>,
+        Error,
+    > {
         let peer = msg.peer;
         let protocols = msg.protocols;
 
-        let (protocol, stream) = self.open_substream(peer, protocols).await?;
+        let (control, _) = self.controls.get(&peer).ok_or(Error::NoConnection(peer))?;
+
+        let fut = {
+            let connection_timeout = self.connection_timeout;
+            let control = control.clone();
+            async move {
+                let (protocol, stream) =
+                    Self::open_substream(control, peer, protocols, connection_timeout).await?;
+
+                Ok((protocol, stream))
+            }
+        };
 
-        Ok((protocol, stream))
+        Ok(Box::pin(fut))
```

Comment:
An alternate solution to returning the future would be to use `Context::join` (used to be `handle_while`), e.g:

```rust
ctx.join(self, fut).await
```

This would let the endpoint handle other messages whilst it is waiting for the future to evaluate.

I am not sure of the implications in terms of how this would look in the trace (how do the spans show up?), overflow behaviour (I think it should not overflow, as these are boxed?) in the case of nested `join`s, or overhead, but it might be a little cleaner.

Reply:
If it nests too deeply, I wonder if `poll` might overflow the stack?

Reply:
I like the suggestion, but if we `await` the `join_future` _within_ the handler, isn't it equivalent to just awaiting the future itself? The way I see it, until we exit the handler (because the `join_future` has completed) we cannot process any other messages in the actor.

Reply:
`join` essentially handles other messages whilst inside of a handler itself and waiting for a given future to complete. In a way, it explicitly breaks the invariant that the actor handles one message at a time - you opt into this by using `join`. 

It might not be suited here, as it was intended to solve more limited, one-time use cases such as circular initialisation messages: https://github.com/Restioson/xtra/issues/13. The way it is used here would cause it to nest as many times as there are incoming protocol messages in the worst case of a 20s timeout being reached.

Reply:
Ahhh, thanks for the explanation. It's really cool that `join` was introduced nevertheless! Somehow I didn't understand from the documentation that it was meant to break that invariant. I think I got it confused with `scoped`.

Reply:
I should probably explicitly word it like that in the docs too!